### ISSUE TITLE
esrecurse uses unmonkeypatched estraverse

### ISSUE THREAD

**Issue by IanVS** [eslint/eslint#5007]
There is a bug when getting `sourceCode` from code containing JSX, and then passing it back into `eslint.verify()`.  The problem seems to be that esrecurse uses a version of estraverse that we have not monkeypatched to include JSX keys.  Because of this, esrecurse visits all properties of the node (https://github.com/estools/esrecurse/blob/master/esrecurse.js#L73). Normally, this isn't a problem when simply linting text, as the esrecurse (called from escope) happens early on in `eslint.verify()` ([here](https://github.com/eslint/eslint/blob/master/lib/eslint.js#L776)), before `parent` properties are added to the AST ([from the enter event](https://github.com/eslint/eslint/blob/master/lib/eslint.js#L822)).  But, after linting is performed and the `sourceCode` is obtained, the AST will now contain the `parent` properties, which causes a stack overflow when passed to `escope.analyze()`.

To solve this, @mysticatea has suggested monkeypatching estraverse similarly to https://github.com/babel/babel-eslint/blob/master/index.js#L38-L56.  I have no real experience with monkeypatching, so I have no idea how we could/should go about this.  But, it's important to solve if we want to allow autoconfig to work on JSX code, because it uses sourceCode extensively (for performance reasons).

**Comment by IanVS** [comment]
@michaelficarra, do you have any thoughts or insights here on how we might go about using escope on a JSX AST?  Is monkeypatching our best option, in your opinion?

**Comment by michaelficarra** [comment]
The problem you're having is caused by esrecurse attempting to operate on unrecognised AST nodes by recursing onto every field in the object. In my opinion, esrecurse should fail immediately when it encounters an unrecognised AST node.

I don't like the monkeypatching of escope (I would prefer a specification be passed in instead), but it seems the best way to move forward given the current state of things.

**Comment by nzakas** [comment]
@IanVS what version are you seeing this in? What is it you're trying to do?

**Comment by IanVS** [comment]
I am seeing this in my [autoconfig branch](https://github.com/eslint/eslint/tree/autoconfig), which admittedly is slightly behind master.

I am trying to lint JSX code, get the `sourceCode`, and re-lint the `sourceCode`.

**Comment by nzakas** [comment]
So it seems like the problem is the `parent` node confuses `esrecurse`?

Also seems like the right answer is for `esrecurse` to expose `Syntax` and `VisitorKeys` directly rather than relying on `estraverse`?

**Comment by mysticatea** [comment]
I think there are 3 approaches.

1. The right approach is that we send PRs to add an option for `visitorKeys` to `esrecurse` and `escope` (the `esrecurse` is used from `escope` internal). It might take long time.
2. We make the `parent` property unenumerable: `Object.defineProperty(node, "parent", {writable: true, configurable: true, enumerable: false})`. But I know this is very slow.
3. We use monkeypatching. It's quick and sloppy.

**Comment by nzakas** [comment]
We can try making `parent` nonenumerable, though we should doubke-check the perf impact (see https://www.nczonline.net/blog/2015/11/performance-implication-object-defineproperty/ )

I think ultimately making such changes to escope/esrecurse make the most sense.

**Comment by mysticatea** [comment]
I checked the performance change.
And I confirmed nonenumerable fixed this bug.

| | Before | After | Rate |
|------|--------|-------|------|
| single file | 2422ms | 2515ms | +4% |
| multi files | 7480ms | 8294ms | +11% |

```diff
@@ -820,7 +820,12 @@ module.exports = (function() {
              */
             controller.traverse(ast, {
                 enter: function(node, parent) {
-                    node.parent = parent;
+                    Object.defineProperty(node, "parent", {
+                        value: parent,
+                        configurable: true,
+                        enumerable: false,
+                        writable: true
+                    });
                     eventGenerator.enterNode(node);
                 },
                 leave: function(node) {
```

```
Before

Single File:
  CPU Speed is 3392 with multiplier 7500000
  Performance Run #1:  2422.559916ms
  Performance Run #2:  2488.714065ms
  Performance Run #3:  2452.773965ms
  Performance Run #4:  2413.956653ms
  Performance Run #5:  2392.814938ms
  Performance budget exceeded: 2422.559916ms (limit: 2211.0849056603774ms)


Multi Files:
  CPU Speed is 3392 with multiplier 24000000
  Performance Run #1:  7571.863146ms
  Performance Run #2:  7449.154811ms
  Performance Run #3:  7428.685312ms
  Performance Run #4:  7480.038359ms
  Performance Run #5:  7559.607224ms
  Performance budget exceeded: 7480.038359ms (limit: 7075.471698113208ms)
```

```
After

Single File:
  CPU Speed is 3392 with multiplier 7500000
  Performance Run #1:  2527.138674ms
  Performance Run #2:  2505.810115ms
  Performance Run #3:  2515.523575ms
  Performance Run #4:  2507.8005ms
  Performance Run #5:  2525.665959ms
  Performance budget exceeded: 2515.523575ms (limit: 2211.0849056603774ms)


Multi Files:
  CPU Speed is 3392 with multiplier 24000000
  Performance Run #1:  8360.738489ms
  Performance Run #2:  8294.080253ms
  Performance Run #3:  8311.994915ms
  Performance Run #4:  8145.785799ms
  Performance Run #5:  8192.124552ms
  Performance budget exceeded: 8294.080253ms (limit: 7075.471698113208ms)
```

```js
it("should not crash when we reuse the SourceCode object", function() {
    eslint.verify("function render() { return <div/> }", { parserOptions: { ecmaVersion: 6, ecmaFeatures: {jsx: true} }});
    eslint.verify(eslint.getSourceCode(), { parserOptions: { ecmaVersion: 6, ecmaFeatures: {jsx: true} }});
});
```

**Comment by mysticatea** [comment]
From curiosity, in the case of monkeypatching.

```diff
diff --git a/lib/util/estraverse.js b/lib/util/estraverse.js
index d2e0d9a..57c4b70 100644
--- a/lib/util/estraverse.js
+++ b/lib/util/estraverse.js
@@ -10,8 +10,10 @@
 // Requirements
 //------------------------------------------------------------------------------

-var estraverse = require("estraverse"),
-    jsxKeys = require("estraverse-fb/keys");
+var path = require("path"),
+    estraverse = require("estraverse"),
+    jsxKeys = require("estraverse-fb/keys"),
+    resolve = require("resolve").sync;

 //------------------------------------------------------------------------------
 // Helers
@@ -48,6 +50,12 @@ installKeys(jsxKeys);
 // Add Experimental node types.
 installKeys(experimentalKeys);

+// Patch for escope.
+var escopePath = resolve("escope", {basedir: __dirname});
+var esrecursePath = resolve("esrecurse", {basedir: path.dirname(escopePath)});
+var estraversePath = resolve("estraverse", {basedir: path.dirname(esrecursePath)});
+require(estraversePath).VisitorKeys = estraverse.VisitorKeys;
+
 //------------------------------------------------------------------------------
 // Public Interface
 //------------------------------------------------------------------------------
```

```
Single File:
  CPU Speed is 3392 with multiplier 7500000
  Performance Run #1:  2406.261347ms
  Performance Run #2:  2421.786582ms
  Performance Run #3:  2409.243905ms
  Performance Run #4:  2462.760597ms
  Performance Run #5:  2407.1527029999997ms
  Performance budget exceeded: 2409.243905ms (limit: 2211.0849056603774ms)


Multi Files:
  CPU Speed is 3392 with multiplier 24000000
  Performance Run #1:  7508.741962ms
  Performance Run #2:  7546.818842ms
  Performance Run #3:  7637.074022ms
  Performance Run #4:  7607.168461ms
  Performance Run #5:  7588.32441ms
  Performance budget exceeded: 7588.32441ms (limit: 7075.471698113208ms)
```

**Comment by IanVS** [comment]
Thanks for the investigations, @mysticatea.  Is there any consensus on how to move forward with this issue?

**Comment by mysticatea** [comment]
IMO (not strongly), applying the monkey patching for right now. Then I will send PRs to esrecurse and escope later.

**Comment by nzakas** [comment]
:-1: to monkey patching. We don't know how that will negatively affect things like babel-eslint. Let's just hold off and do it the right way.

**Comment by mysticatea** [comment]
New version of esrecurse was released.
I'm working on escope.

**Comment by IanVS** [comment]
Way to go, nicely done!

**Comment by mysticatea** [comment]
Just I'm waiting for new release of `escope`.

**Comment by IanVS** [comment]
Awesome, thanks @mysticatea!
